ambry-account/src/main/java/com/github/ambry/account/mysql/DatasetDao.java

```diff
@@ -144,10 +145,14 @@ public DatasetDao(MySqlDataAccessor dataAccessor, MySqlAccountServiceConfig mySq
         String.format("select %s, %s from %s " + "where (%s IS NULL or %s > now(3)) and %s = ? and %s = ? and %s = ?",
             VERSION, DELETE_TS, DATASET_VERSION_TABLE, DELETE_TS, DELETE_TS, ACCOUNT_ID, CONTAINER_ID, DATASET_NAME);
     // list all valid versions sorted by last modified time, and skip the first N records which is not out of retentionCount.
-    listVersionByModifiedTimeSql = String.format("select %s, %s from %s "
+    listVersionByModifiedTimeAndFilterByRetentionSql = String.format("select %s, %s from %s "
             + "where (%s IS NULL or %s > now(3)) and %s = ? and %s = ? and %s = ? and %s = ? ORDER BY %s DESC LIMIT ?, 100",
         VERSION, DELETE_TS, DATASET_VERSION_TABLE, DELETE_TS, DELETE_TS, ACCOUNT_ID, CONTAINER_ID, DATASET_NAME,
         DATASET_VERSION_STATE, LAST_MODIFIED_TIME);
+    listValidDatasetVersionsSql = String.format("select %1$s from %2$s "
+            + "where (%3$s IS NULL or %3$s > now(3)) and %4$s = ? and %5$s = ? and %6$s = ? and %7$s = ? and (%1$s >= ? or ? IS NULL) "
```

Comment:
we should check if the page token is NULL out side of this query statement, we should just do `AND %1$s >= ?`